src/test/kotlin/com/atlassian/performance/tools/hardware/HardwareApdexAnalysisIT.kt

```diff
@@ -0,0 +1,269 @@
+    }
+
+    private fun apdexMetrics(targetFile: File): ApdexResults {
+
+        val run = targetFile.name
+        val workspace = TestWorkspace(targetFile.toPath())
+
+        val cohortResults = targetFile.listFiles()
+            .filter { file ->
+                // I think this is a linear folder structure only ever 1 cohort per run? So potentially this could be simpler
+                val cohortFilter = "**/*$jiraInstanceTypeOfInterest*"
+                matchesSearchPattern(file.absolutePath, cohortFilter)
+            }
+            .map { file ->
+
+                val cohortResult = workspace.readResult(file.name)
+                val failure = cohortResult.failure
+                if (failure == null) {
+                    
+                    val postProcessedResult = processResults(cohortResult)
+
+                    val cohort = postProcessedResult.cohort
+                    if (postProcessedResult.failure != null) {
+                        throw Exception("$cohort failed", postProcessedResult.failure)
+                    }
+
+                    // reproduce the apdex calculation from HardwareExploration
+                    val labels = listOf(
+                        ViewBacklogAction.VIEW_BACKLOG,
+                        VIEW_BOARD,
+                        VIEW_ISSUE,
+                        VIEW_DASHBOARD,
+                        SEARCH_WITH_JQL,
+                        ADD_COMMENT_SUBMIT,
+                        CREATE_ISSUE_SUBMIT,
+                        EDIT_ISSUE_SUBMIT,
+                        PROJECT_SUMMARY,
+                        BROWSE_PROJECTS,
+                        BROWSE_BOARDS
+                    ).map { it.label }
+                    val metrics = postProcessedResult.actionMetrics.filter { it.label in labels }
+                    val apdex = Apdex().score(metrics)
+
+                    // keep the per-action apdex as 'weight' this applied to OK and ERROR actionMetrics as we want to keep the ERRORs on hand for later use
+                    val weightedMetrics = Apdex().weight(metrics)
+
+                    // filter out the ERRORs before re-calculating apdex
+                    val apdexW = Apdex().average(weightedMetrics.filter { it.actionMetric.result == ActionResult.OK })
+
+                    // ensure our new calculation matches the original style
+                    if(apdex != apdexW) {
+                        throw Exception("apdex $apdex != apdexW $apdexW")
+                    }
+
+                    ApdexResults(run.toInt(), apdex, weightedMetrics)
+
+                } else {
+                    println("ignominious failure! ${failure.localizedMessage}")
+                    BugAwareTolerance(logger).handle(failure, workspace)
+                    null
+                }
+            }
+            // this seems very hacky, I should just be able to get 1 result out of the sequence code above ... ?
+            .first{ it is ApdexResults}
```

Comment:
This tries to detect whether we `map`ped to an `ApdexResults` or a `null`.
We can do it directly, by changing the `map` above to a `mapNotNull` and then `first()`.

PS. We can extract the `map` contents to a private method to see the full chain at once.

Reply:
done 👍